### DISCUSSION TITLE
13. September 2022

### DISCUSSION THREAD

**Discussion by dimakuv** [gramineproject/gramine#892] [Meeting notes]
### Agenda

*(please write your proposed agenda items in comments under this discussion)*

- [x] *Mona/Wojtek*: download statistics
- [x] *Mona*: how to respond to queries at [[gramine-users@googlegroups.com](mailto:gramine-users@googlegroups.com)] and should we look at expanding that list to be beyond maintainers to get some additional support
- [x] *Vijay*: PAL VMA booking removal + EDMM integration
- [ ] *Sankar/Dmitrii*: Requirement for use of HSM (including MS KeyVault) where all keys are stored (including Enclave Signing Key and Master Encryption Key), and these keys must never be seen in the clear:
    - Enclave Signing Key: use the special "Signing API" of HSM on half-baked SIGSTRUCT
    - Master Encryption Key: write a tool that runs in Gramine-SGX and uses the special "Wrap-Key API" of HSM to bring the Encryption Key inside of the enclave securely
- [ ] *Anees/Dmitrii*: Basic security of the enclave signing key: protection with a password?
    - May be helpful: https://cryptography.io/en/latest/search/?q=password&check_keywords=yes&area=default#
    - https://stackoverflow.com/questions/4294689/how-to-generate-an-openssl-key-using-a-passphrase-from-the-command-line

### Download statistics

We want to have some download statistics, just to have a general idea of how popular Gramine is and what are the trends.

Woju opposed the idea to install GitHub plugins. But he said he could do smth like what they do for Qubes: https://www.qubes-os.org/statistics/.

*Woju*: there are problems with GitHub default counting approaches, they break privacy. Qubes has an "estimated package downloads" graph (actually, IP addresses that downloaded the packages). The detailed log is never released and is periodically cleaned. But if people are behind a corporate proxy (with a single IP for thousands of employees), then the diagram is very much skewed. So the vertical bar is not really accurate, but the trends and ratio of releases reflect reality. Two problems: Dockerhub downloads are not tracked (but Dockerhub has smth called the "Pulls" number), and once Gramine is included in OS repos, also can't count those downloads.

*Woju*: changes that we need:
- we must have a web server that logs each download, from this log we parse the release version and date/time;
- then feed this log into Woju's script.

**TODO(Woju)**: will look into how to implement this for our gramineproject.io website.

### How to respond to queries at gramine-users@googlegroups.com

*Mona*: The traffic increased in the mailing list. We need to have *timely* replies.

*Dmitrii*: maybe redirect people from mailing lists to GitHub Discussions?

*Michal* will create a form that asks users do they want to create an Issue or to create a Discussion. This will effectively move people away from mailing lists to GitHub.

*Mona*: let's update our pages to direct users not to mailing lists, but to GitHub Discussions.

**TODO(Michal)**: update.

### PAL VMA booking removal + EDMM integration

Two cases:
1. Changing page permissions (need to keep previous permissions on the page).
2. `PalVirtualMemoryAlloc` function -- it only has current requested permissions. It is possible to have overlapping regions, and for it we have different cases: for new region, everything as normal; for overlapping region, somehow need to pass the previous permissions (for those already-allocated pages).

*Borys*: no clear solution for case 2. Allocating memory is a two-step process: first bookkeep, then call PAL API. Currently if we have overlapping alloc, then it is PAL's job -- and in Linux PAL the host does everything, in Linux SGXv1 PAL there is nothing to do.

*Don*: the LibOS notices overlapping regions, and splits one PAL `PalVirtualMemoryAlloc` call into several calls. *Borys*: this sounds complicated, e.g. we need to put all this under one lock acquisition, but `PalVirtualMemoryAlloc` itself can ask LibOS for more memory allocation (Dmitrii doesn't see why this could be needed). *Borys*: one more problem, LibOS bookkeeping itself might need memory, so it allocates some internal object, so it bookkeeps this memory.

**TODO(Vijay)**: implement the idea of LibOS only giving non-overlapping memory regions to PAL. So, for overlapped region LibOS must issue: PAL unmap + PAL map.

**Comment by woju** [comment]
- [ ] Mona/Wojtek: download statistics, about 15 min

**Comment by monavij** [comment]
Discuss how to respond to queries at [gramine-users@googlegroups.com] and should we look at expanding that list to be beyond maintainers to get some additional support.
